File: admin_panel/lib/api/lessons.ts
import { apiFetch, getApiConfig } from "@/lib/api/config";

export interface LessonItem {
  id: string;
  course_id: string;
  section_id: string | null;
  title: string;
  video_url: string;
  order_no: number;
  is_free: boolean;
}

function headers() {
  const { adminApiKey } = getApiConfig();
  return {
    "Content-Type": "application/json",
    ...(adminApiKey ? { "x-admin-api-key": adminApiKey } : {}),
  };
}

async function parseError(response: Response, fallback: string) {
  try {
    const data = (await response.json()) as { detail?: string };
    return data.detail?.trim() || fallback;
  } catch {
    return fallback;
  }
}

export async function fetchLessons(courseId?: string) {
  const suffix = courseId ? `?course_id=${encodeURIComponent(courseId)}` : "";
  const response = await apiFetch(`/api/v1/lessons${suffix}`, { cache: "no-store" });
  if (!response.ok) throw new Error(await parseError(response, "Darslarni olishda xatolik."));
  const data = (await response.json()) as { items: LessonItem[] };
  return data.items ?? [];
}

export async function createLesson(payload: {
  course_id: string;
  title: string;
  video_url: string;
  order_no: number;
  is_free: boolean;
  section_id?: string | null;
}) {
  const response = await apiFetch("/api/v1/lessons", {
    method: "POST",
    headers: headers(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseError(response, "Dars qo'shilmadi."));
  return (await response.json()) as LessonItem;
}

export async function updateLesson(id: string, payload: Partial<Omit<LessonItem, "id">>) {
  const response = await apiFetch(`/api/v1/lessons/${id}`, {
    method: "PATCH",
    headers: headers(),
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseError(response, "Dars yangilanmadi."));
  return (await response.json()) as LessonItem;
}

export async function removeLesson(id: string) {
  const response = await apiFetch(`/api/v1/lessons/${id}`, {
    method: "DELETE",
    headers: headers(),
  });
  if (!response.ok) throw new Error(await parseError(response, "Dars o'chirilmadi."));
}

File: fastapi/app/schemas/lessons.py
from datetime import datetime

from pydantic import BaseModel, Field


class LessonCreate(BaseModel):
    course_id: str = Field(min_length=1)
    section_id: str | None = None
    title: str = Field(min_length=1, max_length=300)
    video_url: str = ""
    order_no: int = Field(default=1, ge=1)
    is_free: bool = False


class LessonUpdate(BaseModel):
    course_id: str | None = None
    section_id: str | None = None
    title: str | None = None
    video_url: str | None = None
    order_no: int | None = Field(default=None, ge=1)
    is_free: bool | None = None


class LessonItem(BaseModel):
    id: str
    course_id: str
    section_id: str | None
    title: str
    video_url: str
    order_no: int
    is_free: bool
    created_at: datetime


class LessonListResponse(BaseModel):
    items: list[LessonItem]
